### ISSUE TITLE
Possible unit_arg false positive when passing `Default::default()` as function parameter resolving to unit value

### ISSUE THREAD

**Issue by Enet4** [rust-lang/rust-clippy#7341]
Lint name: `unit_arg`

I tried this code ([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=750e3a4fb51d62ed3034e5bc53d37779)):

```rust
#[derive(Debug)]
struct Foo;

trait Factory: Sized {
    type Options: Default;

    fn create(options: Self::Options) -> Self;
}

impl Factory for Foo {
    type Options = ();

    fn create(_: Self::Options) -> Self {
        Foo
    }
}

fn main() {
    let x = Foo::create(Default::default());

    println!("{:?}", x);
}
```

In my real use case, the same pattern is employed for generalizing the passing of default options to a creator function, irrespective of implementation.

I expected this not to trigger `unit_arg`, and most importantly, not to recommend moving `Default::default();` to the front while leaving a literal unit value in its stead, which could be incompatible with other implementations of `Factory`:

```none
error: passing a unit value to a function
  --> src/main.rs:21:13
   |
21 |     let x = Foo::create(Default::default());
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![deny(clippy::unit_arg)]
   |         ^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unit_arg
help: move the expression in front of the call and replace it with the unit literal `()`
   |
21 |     let x = {
22 |         Default::default();
23 |         Foo::create(())
24 |     };
   |
```

### Meta

- `cargo clippy -V`: clippy 0.1.52 (9bc8c42 2021-05-09), also occurs in 0.1.54 (2021-06-09 eab201d)
- `rustc -Vv`:
  ```
  rustc 1.52.1 (9bc8c42bb 2021-05-09)
  binary: rustc
  commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
  commit-date: 2021-05-09
  host: x86_64-unknown-linux-gnu
  release: 1.52.1
  LLVM version: 12.0.0
  ```

**Comment by camsteffen** [comment]
It seems reasonable to me to suggest `Foo::create(())` here since the type of `Options` is known at compile time.

As for re-locating `Default::default()`, we should remove that part of the suggestion for known types like unit, any primitive, and maybe any std type.
